File: PortScanner.java
// PortScanner.java
import com.beust.jcommander.JCommander;
import com.beust.jcommander.Parameter;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.io.*;
import java.net.*;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.util.concurrent.*;

public class PortScanner {
    @Parameter(names = "--host", required = true)
    private String host;
    @Parameter(names = "--ports")
    private String ports = "22,80,443,1-1024";
    @Parameter(names = "--timeout")
    private int timeout = 2;
    @Parameter(names = "--threads")
    private int threads = 10;
    @Parameter(names = "--json")
    private String jsonFile;
    @Parameter(names = "--csv")
    private String csvFile;
    @Parameter(names = "--verbose")
    private boolean verbose;
    @Parameter(names = "--no-color")
    private boolean noColor;

    private InetAddress ip;
    private List<Integer> portList;
    private List<ScanResult> results = Collections.synchronizedList(new ArrayList<>());
    private ExecutorService executor;
    private boolean color;

    static class ScanResult {
        int port;
        String state;
        String banner;
    }

    public void run() throws Exception {
        ip = InetAddress.getByName(host);
        portList = parsePorts(ports);
        threads = Math.min(threads, portList.size());
        color = !noColor && System.console() != null;
        executor = Executors.newFixedThreadPool(threads);

        System.out.println("Scanning " + host + " (" + ip.getHostAddress() + ")...");
        long start = System.currentTimeMillis();

        List<Future<ScanResult>> futures = new ArrayList<>();
        for (int port : portList) {
            futures.add(executor.submit(() -> scanPort(port)));
        }
        for (Future<ScanResult> f : futures) {
            ScanResult res = f.get();
            results.add(res);
            if (verbose) printResult(res);
        }
        executor.shutdown();
        long elapsed = (System.currentTimeMillis() - start) / 1000.0;
        long openCount = results.stream().filter(r -> "open".equals(r.state)).count();
        System.out.printf("\nScan completed in %.2fs. Found %d open ports.\n", elapsed, openCount);

        if (jsonFile != null) exportJSON();
        if (csvFile != null) exportCSV();
    }

    private List<Integer> parsePorts(String s) {
        Set<Integer> set = new HashSet<>();
        for (String part : s.split(",")) {
            if (part.contains("-")) {
                String[] range = part.split("-");
                int start = Integer.parseInt(range[0]);
                int end = Integer.parseInt(range[1]);
                for (int p = start; p <= end; p++) set.add(p);
            } else {
                set.add(Integer.parseInt(part));
            }
        }
        return new ArrayList<>(set);
    }

    private ScanResult scanPort(int port) {
        ScanResult res = new ScanResult();
        res.port = port;
        try (Socket socket = new Socket()) {
            socket.connect(new InetSocketAddress(ip, port), timeout * 1000);
            res.state = "open";
            res.banner = "";
        } catch (SocketTimeoutException e) {
            res.state = "filtered";
            res.banner = "";
        } catch (ConnectException e) {
            if (e.getMessage().contains("refused")) {
                res.state = "closed";
                res.banner = "";
            } else {
                res.state = "error";
                res.banner = e.getMessage();
            }
        } catch (Exception e) {
            res.state = "error";
            res.banner = e.getMessage();
        }
        return res;
    }

    private void printResult(ScanResult res) {
        if (color) {
            String colorCode = res.state.equals("open") ? "\u001B[32m" : res.state.equals("closed") ? "\u001B[31m" : res.state.equals("filtered") ? "\u001B[33m" : "\u001B[37m";
            System.out.printf("%sPort %d/tcp  %s\033[0m  %s%n", colorCode, res.port, res.state, res.banner);
        } else {
            System.out.printf("Port %d/tcp  %s  %s%n", res.port, res.state, res.banner);
        }
    }

    private void exportJSON() throws IOException {
        Map<String, Object> data = new HashMap<>();
        data.put("host", host);
        data.put("ip", ip.getHostAddress());
        data.put("results", results);
        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        String json = gson.toJson(data);
        try (FileWriter fw = new FileWriter(jsonFile)) {
            fw.write(json);
        }
        System.out.println("Results exported to " + jsonFile);
    }

    private void exportCSV() throws IOException {
        try (PrintWriter pw = new PrintWriter(new FileWriter(csvFile))) {
            pw.println("port,state,banner");
            for (ScanResult r : results) {
                pw.printf("%d,%s,\"%s\"%n", r.port, r.state, r.banner);
            }
        }
        System.out.println("Results exported to " + csvFile);
    }

    public static void main(String[] args) throws Exception {
        PortScanner scanner = new PortScanner();
        JCommander.newBuilder().addObject(scanner).build().parse(args);
        scanner.run();
    }
}

File: PortScanner.kt
// PortScanner.kt
import com.beust.jcommander.JCommander
import com.beust.jcommander.Parameter
import com.google.gson.GsonBuilder
import java.io.File
import java.net.*
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

data class ScanResult(val port: Int, val state: String, val banner: String = "")

class PortScanner {
    @Parameter(names = ["--host"], required = true)
    private lateinit var host: String

    @Parameter(names = ["--ports"])
    private var portsSpec: String = "22,80,443,1-1024"

    @Parameter(names = ["--timeout"])
    private var timeout: Int = 2

    @Parameter(names = ["--threads"])
    private var threads: Int = 10

    @Parameter(names = ["--json"])
    private var jsonFile: String? = null

    @Parameter(names = ["--csv"])
    private var csvFile: String? = null

    @Parameter(names = ["--verbose"])
    private var verbose: Boolean = false

    @Parameter(names = ["--no-color"])
    private var noColor: Boolean = false

    private lateinit var ip: InetAddress
    private lateinit var ports: List<Int>
    private val results = ConcurrentLinkedQueue<ScanResult>()
    private var color: Boolean = false

    fun run() {
        ip = InetAddress.getByName(host)
        ports = parsePorts(portsSpec)
        threads = minOf(threads, ports.size)
        color = !noColor && System.console() != null

        println("Scanning $host (${ip.hostAddress})...")
        val start = System.currentTimeMillis()

        val executor = Executors.newFixedThreadPool(threads)
        val futures = ports.map { port ->
            executor.submit<ScanResult> { scanPort(port) }
        }
        futures.forEach { future ->
            val res = future.get()
            results.add(res)
            if (verbose) printResult(res)
        }
        executor.shutdown()
        executor.awaitTermination(1, TimeUnit.MINUTES)

        val elapsed = (System.currentTimeMillis() - start) / 1000.0
        val openCount = results.count { it.state == "open" }
        println("\nScan completed in ${String.format("%.2f", elapsed)}s. Found $openCount open ports.")

        jsonFile?.let { exportJSON(it) }
        csvFile?.let { exportCSV(it) }
    }

    private fun parsePorts(s: String): List<Int> {
        val set = mutableSetOf<Int>()
        s.split(',').forEach { part ->
            if (part.contains('-')) {
                val (start, end) = part.split('-').map { it.toInt() }
                (start..end).forEach { set.add(it) }
            } else {
                set.add(part.toInt())
            }
        }
        return set.toList()
    }

    private fun scanPort(port: Int): ScanResult {
        return try {
            val socket = Socket()
            socket.soTimeout = timeout * 1000
            val addr = InetSocketAddress(ip, port)
            socket.connect(addr, timeout * 1000)
            socket.close()
            ScanResult(port, "open")
        } catch (e: SocketTimeoutException) {
            ScanResult(port, "filtered")
        } catch (e: ConnectException) {
            if (e.message?.contains("refused") == true) {
                ScanResult(port, "closed")
            } else {
                ScanResult(port, "error", e.message ?: "")
            }
        } catch (e: Exception) {
            ScanResult(port, "error", e.message ?: "")
        }
    }

    private fun printResult(res: ScanResult) {
        if (color) {
            val colorCode = when (res.state) {
                "open" -> "\u001B[32m"
                "closed" -> "\u001B[31m"
                "filtered" -> "\u001B[33m"
                else -> "\u001B[37m"
            }
            println("$colorCodePort ${res.port}/tcp  ${res.state}\u001B[0m  ${res.banner}")
        } else {
            println("Port ${res.port}/tcp  ${res.state}  ${res.banner}")
        }
    }

    private fun exportJSON(filename: String) {
        val data = mapOf(
            "host" to host,
            "ip" to ip.hostAddress,
            "results" to results.toList()
        )
        val gson = GsonBuilder().setPrettyPrinting().create()
        File(filename).writeText(gson.toJson(data))
        println("Results exported to $filename")
    }

    private fun exportCSV(filename: String) {
        File(filename).printWriter().use { pw ->
            pw.println("port,state,banner")
            results.forEach { pw.println("${it.port},${it.state},\"${it.banner}\"") }
        }
        println("Results exported to $filename")
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val scanner = PortScanner()
            JCommander.newBuilder().addObject(scanner).build().parse(*args)
            scanner.run()
        }
    }
}
